File: python/test_sort.py

import time, sys
import argparse

def load(infile):
    a = []
    tstart = time.time()
    for line in open(infile):
        line = line.strip()
        if len(line) == 0: continue
        a.append(int(line))
    
    print 'load time spend %f s' % (time.time() - tstart)
    print 'array size %d' % len(a)
    
    return a

def quickSort(a, ibegin, iend):
    if ibegin >= iend: return
    tmp = a[ibegin]
    low, high = ibegin, iend
    while low != high:
        while low != high and a[high] >= tmp:
            high = high - 1
        if low != high: a[high] = a[low]
        while low != high and a[low] <= tmp:
            low = low + 1
        if low != high: a[low] = a[high]
    a[low] = tmp
    quickSort(a, ibegin, low - 1)
    quickSort(a, low + 1, iend)

if __name__ == '__main__':
    parser = argparse.ArgumentParser()
    parser.add_argument('infile', help = 'random number file')
    args = parser.parse_args()
    a = load(args.infile)
    if len(a) <= 20:
        print a
    
    b = a[:]
    
    tstart = time.time()
    a.sort()
    print 'list.sort spend %f s' % (time.time() - tstart)
    if len(a) <= 20:
        print a
    
    a = b[:]
    tstart = time.time()
    quickSort(a, 0, len(a) - 1)
    print 'quick sort spend %f s' % (time.time() - tstart)
    if len(a) <= 20:
        print b
#         print a

File: lua/test_sort.lua

require 'xutils'
require 'torch'

local function runSort(arr, sort_alg, sort_alg_lbl)
	local tmp_arr = xcopya(arr)
	print(string.format('==%s demo==', sort_alg_lbl))
	print 'before sort'
	if #arr <= 20 then
		xprinta(arr)
	end
	local tstart = os.clock()
	sort_alg(arr)
	printf('time spend %f s\n', os.clock() - tstart)
	print 'after sort'
	if #arr <= 20 then
		xprinta(arr)
	end
	print(string.format('==%s demo done==\n', sort_alg_lbl))
	xcopytoa(arr, tmp_arr)
	tmp_arr = nil
end

local function loadRandNumbers(infile)
	local tstart = os.clock()
	local fin = io.open(infile, 'r')
	local line = nil
	local arr = {}
	while true do
		line = fin:read()
		if not line then break end
		table.insert(arr, tonumber(line))
	end
	fin:close()
	printf('load time %f s\n', os.clock() - tstart)

	printf('arr size = %d\n', #arr)

	return arr
end

local function loadRandNumbers2(infile)
	local tstart = os.clock()
	local fin = io.open(infile, 'r')
	local line = nil
	local arr = {}
	while true do
		line = fin:read('*number')
		if not line then break end
		table.insert(arr, line)
	end
	fin:close()
	printf('load time %f s\n', os.clock() - tstart)

	printf('arr size = %d\n', #arr)

	return arr
end

local function getArgs()
	local cmd = torch.CmdLine()
	cmd:text('test Lua native sort algorithm and qsort I implmented')
	cmd:option('-infile', '', 'input number file')
	
	return cmd:parse(arg)
end

local function main()
	local opts = getArgs()
	local arr = loadRandNumbers2(opts.infile)
	runSort(arr, table.sort, 'native sort')
	runSort(arr, xquick_sort, 'quick sort')
end

main()

